include/uapi/sound/sof/header.h

```diff
@@ -26,6 +26,30 @@ struct sof_abi_hdr {
 	__u32 data[];		/**< Component data - opaque to core */
 }  __packed;
 
+/*
+ * Identifies data type, size and ABI for IPC4.
+ * Used by any bespoke component data structures or binary blobs.
+ */
+struct sof_ipc4_abi_hdr {
+	__u32 magic;		/**< 'S', 'O', 'F', '4' */
+	__u32 rsvd1;
+	__u32 size;		/**< size in bytes of data excl. this struct */
+	__u32 abi;		/**< SOF ABI version */
```

Comment:
what's the definition for this ABI in the IPC4 context?

Reply:
The header is the identical to ipc3 but we taken into use two u32 from the reserved section, right?
Do we really need a new struct for the same thing?
Can we reuse the `type` for example?

Would something like this worth to consider:
```c
/**
 * struct sof_abi_hdr - Header for all non IPC ABI data.
 * @magic: for IPC3: 'S', 'O', 'F', '\0'
 * 	   for IPC4: 'S', 'O', 'F', '4'
 * @type: IPC3: component specific type
 * 	  IPC4: IPC type of the blob: 0: init, 2: config, 4: large_config
 * @abi: SOF ABI version
 * @ipc4.extension: Extension value in IPC message header
 * @reserved: reserved for future use
 * @data: Component data - opaque to core
 *
 * Identifies data type, size and ABI.
 * Used by any bespoke component data structures or binary blobs.
 */
struct sof_abi_hdr {
	__u32 magic;
	__u32 type;
	__u32 size;
	__u32 abi;
	union {
		struct {
			__u32 extension;
		} ipc4;
		__u32 reserved[4];
	};
	__u32 data[];
}  __packed;
```
one can access to extension:
```c
	struct sof_abi_hdr *data = cdata->data;
	struct sof_ipc4_msg msg

	msg.extension = data->ipc4.extension;
```

Reply:
@ujfalusi I would like not to change the old ABI header. Re-use the old one or use a new one are both OK to me. @ranj063 suggested using a new struct for IPC4. What's your opinion? @plbossart @ranj063

Reply:
@libinyang, if we are introducing a new struct for each new member then what is the point to have the reserved[] section in the first place?
A new member in the struct means that the ABI version must be different, so the fix part is always
```c
struct sof_abi_hdr_fixed {
struct sof_abi_hdr {
	__u32 magic;
	__u32 type;
	__u32 size;
	__u32 abi;
	__u8 rest[];
};
```
The depending on the abi version and IPC4 version you would cast different struct to the `rest` in essence.
If you define the exact sized header for the new IPC with the only difference is in the previously reserved area then I don't see why you can not update the struct as it will remain the same with the exception of including the new fields.

Reply:
@libinyang Maybe what we can do if like @ujfalusi suggested keep the original sof_abi_hdr and add a separate struct sof_ipc4_hdr as part of the data[]. Then, the IPC4 byte control handle will need to strip out the top-level header and then the inner ipc4-specific hdr. sof-ctl will need to change accordingly

Reply:
@ujfalusi @ranj063 OK, I will reuse the struct sof_abi_hdr. Thanks for the suggestion.

Reply:
no, sorry @ujfalusi we can't have a union between something and reserved. With all due respect this is not going to work. Either something is allocated or it's reserved, it can't be both.